README.rst

ezdxf
=====

Abstract
--------

A Python package to create and modify DXF drawings, independent from the DXF
version. You can open/save every DXF file without losing any content (except comments),
Unknown tags in the DXF file will be ignored but preserved for saving. With this behavior
it is possible to open also DXF drawings that contains data from 3rd party applications.

Quick-Info
----------

- ezdxf is a Python package to create new DXF files and read/modify/write existing DXF files
- the intended audience are developers
- requires at least CPython 3.5, support for Python 2 will be dropped with ezdxf v0.9
- OS independent
- additional required packages: `pyparsing <https://pypi.org/project/pyparsing/>`_
- MIT-License
- read/write/new support for DXF versions: R12, R2000, R2004, R2007, R2010, R2013 and R2018
- additional read support for DXF versions R13/R14 (upgraded to R2000)
- additional read support for older DXF versions than R12 (upgraded to R12)
- preserves third-party DXF content
- additional fast DXF R12 writer, that creates just an ENTITIES section with support for the basic DXF entities

a simple example::

    import ezdxf

    drawing = ezdxf.new(dxfversion='AC1024')  # or use the AutoCAD release name ezdxf.new(dxfversion='R2010')
    modelspace = drawing.modelspace()
    modelspace.add_line((0, 0), (10, 0), dxfattribs={'color': 7})
    drawing.layers.new('TEXTLAYER', dxfattribs={'color': 2})
    # use set_pos() for proper TEXT alignment - the relations between halign, valign, insert and align_point are tricky.
    modelspace.add_text('Test', dxfattribs={'layer': 'TEXTLAYER'}).set_pos((0, 0.2), align='CENTER')
    drawing.saveas('test.dxf')

example for the *r12writer*, writes a simple DXF R12 file without in-memory structures::

    from random import random
    from ezdxf.r12writer import r12writer

    MAX_X_COORD = 1000.0
    MAX_Y_COORD = 1000.0
    CIRCLE_COUNT = 100000

    with r12writer("many_circles.dxf") as dxf:
        for i in range(CIRCLE_COUNT):
            dxf.add_circle((MAX_X_COORD*random(), MAX_Y_COORD*random()), radius=2)

The r12writer supports only the ENTITIES section of a DXF R12 drawing, no HEADER, TABLES or BLOCKS section is
present, except FIXED-TABLES are written, than some additional predefined text styles and line types are available.

Installation
============

Install with pip::

    pip install ezdxf

Install develop version (only if you have to)::

    pip install git+https://github.com/mozman/ezdxf.git@develop

or from source::

    python setup.py install

Website
=======

https://ezdxf.mozman.at/

Documentation
=============

Documentation of development version at https://ezdxf.mozman.at/docs

Documentation of latest release at http://ezdxf.readthedocs.io/

Contribution
============

The source code of ezdxf can be found at GitHub.com:

http://github.com/mozman/ezdxf.git

Only pull requests for the **develop** branch will be accepted.

Feedback
========

Issue Tracker at:

http://github.com/mozman/ezdxf/issues

Questions and Feedback at Google Groups:

https://groups.google.com/d/forum/python-ezdxf

python-ezdxf@googlegroups.com

Feedback is greatly appreciated.

Manfred

Contact
=======

ezdxf@mozman.at


NEWS.rst

News
====

Latest versions of ezdxf (>= 0.9) are not Python 2 compatible

Version 0.8.10 - 2019-02-24
---------------------------

- BUGFIX release for Python 2 compatible main version 0.8
- renamed groups.py into dxfgroups.py to avoid confilicts with cython
- added encoding to files containing utf-8 characters (chr15m)
- fixed some issues when running on linux
- BUGFIX: closed LWPOLYLINE did not work with AutoCAD (order matters), introduced with v0.8.9 packed data

Version 0.8.9 - 2018-11-28
--------------------------

- Release notes: https://ezdxf.mozman.at/release-v0-8-9.html
- IMPORTANT: Python 2 support will be dropped in ezdxf v0.9.0, because Python 2 support get more and more annoying.
- CHANGE: refactoring of internal tag representation for a smaller memory footprint, but with some speed penalty
- NEW: packed data for LWPOLYLINE points, faster __getitem__;  added __setitem__, __delitem__, insert() and append()
  methods; renamed discard_points() in clear(); removed get_rstrip_points() and ctx manager rstrip_points();
  user defined point format;
- NEW: packed data for SPLINE, knots, weights, fit- and control points are stored as array.array();
  Spline.get_knot_values(), Spline.get_weights(), Spline.get_control_points() and Spline.get_fit_points() are deprecated,
  direct access to this attributes by Spline.knot_values, Spline.weights, Spline.control_points and Spline.fit_points,
  all attributes with a list-like interface. Knot, control point and fit point counter updated automatically,
  therefore counters are read only now.
- NEW: packed data for MESH, vertices, faces, edges and edge crease values stored as array.array(), high level interface unchanged
- NEW: Drawing.layouts_and_blocks(), iterate over all layouts (mode space and paper space) and all block definitions.
- NEW: Drawing.chain_layouts_and_blocks(), chain entity spaces of all layouts and blocks. Yields an iterator for all
  entities in all layouts and blocks
- NEW: Drawing.query(), entity query over all layouts and blocks
- NEW: Drawing.groupby(), groups DXF entities of all layouts and blocks by an DXF attribute or a key function
- NEW: Layout.set_redraw_order() and Layout.get_redraw_order(), to change redraw order of entities in model space and
  paper space layouts
- NEW: BlockLayout.is_layout_block, True if block is a model space or paper space block definition
- NEW: ezdxf.algebra.Arc helper class to create arcs from 2 points and an angle or radius, or from 3 points
- NEW: ezdxf.algebra.Arc.add_to_layout() with UCS support to create 3D arcs
- NEW: rename paper space layouts by Drawing.layouts.rename(old_name, new_name)
- NEW: Basic support for embedded objects (new in AutoCAD 2018), ezdxf reads and writes the embedded data as it is,
  no interpretation no modification, just enough to not break DXF files with embedded objects at saving.
- CHANGE: Drawing.blocks.delete_block(name, safe=True), new parameter save, check if block is still referenced
  (raises DXFValueError)
- CHANGE: Drawing.blocks.delete_all_blocks(safe=True), if parameter safe is True, do not delete blocks that are still referenced
- BUGFIX: invalid CLASS definition for DXF version R2000 (AC1015) fixed, bug was only triggered at upgrading from R13/R14 to R2000
- BUGFIX: fixed broken Viewport.AcDbViewport property
- `Basic` read support for many missing DXF entities/objects

    - ACAD_PROXY_GRAPHIC
    - HELIX
    - LEADER
    - LIGHT
    - MLEADER (incomplete)
    - MLINE (incomplete)
    - OLEFRAME
    - OLE2FRAME
    - SECTION
    - TABLE (incomplete)
    - TOLERANCE
    - WIPEOUT
    - ACAD_PROXY_OBJECT
    - DATATABLE
    - DICTIONARYVAR
    - DIMASSOC
    - FIELD (incomplete)
    - FIELDLIST (not documented by Autodesk)
    - IDBUFFER
    - LAYER_FILTER
    - MATERIAL
    - MLEADERSTYLE
    - MLINESTYLE
    - SORTENTSTABLE
    - SUN
    - SUNSTUDY (incomplete) (no real world DXF files with SUNSTUDY for testing available)
    - TABLESTYLE (incomplete)
    - VBA_PROJECT (no real world DXF files with embedded VBA for testing available)
    - VISUALSTYLE
    - WIPEOUTVARIABLES
    - for all unsupported entities/objects exist only raw DXF tag support

Version 0.8.8 - 2018-04-02
--------------------------

- Release notes: https://ezdxf.mozman.at/release-v0-8-8.html
- NEW: read/write support for GEODATA entity
- NEW: read/(limited)write support for SURFACE, EXTRUDEDSURFACE, REVOLVEDSURFACE, LOFTEDSURFACE and SWEPTSURFACE entity
- NEW: support for extension dictionaries
- NEW: add_spline_control_frame(), create and add B-spline control frame from fit points
- NEW: add_spline_approx(), approximate B-spline by a reduced count of control points
- NEW: ezdxf.setup_linetypes(dwg), setup standard line types
- NEW: ezdxf.setup_styles(dwg), setup standard text styles
- NEW: LWPolyline.vertices() yields all points as (x, y) tuples in OCS, LWPolyline.dxf.elevation is the z-axis value
- NEW: LWPolyline.vertices_in_wcs() yields all points as (x, y, z) tuples in WCS
- NEW: basic __str__()  and __repr__() support for DXF entities, returns just DXF type and handle
- NEW: bulge related function in module ezdxf.algebra.bulge
- NEW: Object Coordinate System support by DXFEntity.ocs() and OCS() class in module ezdxf.algebra
- NEW: User Coordinate System support by UCS() class in module ezdxf.algebra
- CHANGE: DXFEntity.set_app_data() and Entity.set_xdata accept also list of tuples as tags, DXFTag() is not required
- BUGFIX: entity structure validator excepts group code >= 1000 before XDATA section (used in AutoCAD Civil 3D and AutoCAD Map 3D)

Version 0.8.7 - 2018-03-04
--------------------------

- Release notes: https://ezdxf.mozman.at/release-v0-8-7.html
- NEW: entity.get_layout() returns layout in which entity resides or None if unassigned
- NEW: copy any DXF entity by entity.copy() without associated layout, add copy to any layout you want, by
  layout.add_entity().
- NEW: copy entity to another layout by entity.copy_to_layout(layout)

tests/test_blockssection.py
# Created: 14.03.2011, 2018 rewritten for pytest
# Copyright (C) 2011-2018, Manfred Moitzi
# License: MIT License

from __future__ import unicode_literals
import pytest
from io import StringIO
import sys

import ezdxf
from ezdxf.tools.test import compile_tags_without_handles, load_section
from ezdxf.sections.blocks import BlocksSection
from ezdxf.lldxf.tagwriter import TagWriter
from ezdxf.tools.c23 import PY3


@pytest.fixture(scope='module')
def ac1009():
    return ezdxf.new('AC1009')


@pytest.fixture
def blocks(ac1009):
    return BlocksSection(load_section(TESTBLOCKS, 'BLOCKS', ac1009.entitydb), ac1009)


@pytest.mark.skipif(not PY3 and sys.platform.startswith('linux'), reason="I don't know")
def test_write(blocks):
    stream = StringIO()
    blocks.write(TagWriter(stream))
    result = stream.getvalue()
    stream.close()
    t1 = list(compile_tags_without_handles(TESTBLOCKS))
    t2 = list(compile_tags_without_handles(result))
    assert t1 == t2


def test_empty_section(ac1009):
    blocks = BlocksSection(load_section(EMPTYSEC, 'BLOCKS'), ac1009)
    stream = StringIO()
    blocks.write(TagWriter(stream))
    result = stream.getvalue()
    stream.close()
    assert EMPTYSEC == result


def test_key(blocks):
    assert blocks.key('Test') == 'test'
    block = blocks.new('TEST')
    assert blocks.key(block) == 'test'


def test_create_block(blocks):
    block = blocks.new('TEST')
    assert block in blocks


def test_is_layout_block(blocks):
    block = blocks.new('TEST')
    assert block.is_layout_block is False
    msp = blocks.new('*Model_Space')
    assert msp.is_layout_block is True
    msp = blocks.new('$Model_Space')
    assert msp.is_layout_block is True
    psp = blocks.new('*Paper_Space')
    assert psp.is_layout_block is True
    psp = blocks.new('$Paper_Space')
    assert psp.is_layout_block is True


def test_overwrite_existing_block(blocks):
    block = blocks.new('TEST')
    assert block in blocks
    old_len = len(blocks)
    blocks.new('Test')  # block names are case insensitive
    assert len(blocks) == old_len, 'should overwrite block "TEST"'


def test_not_in_blocks_section(blocks):
    assert 'TEST' not in blocks


def test_getitem(blocks):
    blocks.new('TEST')
    block = blocks['TEST']
    assert 'TEST' == block.name
    block = blocks['Test']
    assert 'TEST' == block.name


def test_case_insensitivity(blocks):
    blocks.new('TEST')
    assert 'TEST' in blocks
    assert 'Test' in blocks


def test_iter_blocks(blocks):
    blocks = list(blocks)
    assert 4 == len(blocks)


def test_block_content_entity_drawing_attribute(blocks, ac1009):
    archtick = blocks['_ARCHTICK']
    entities = list(archtick)
    assert 1 == len(entities)  # VERTEX & SEQEND doesn't count
    e = entities[0]
    assert ac1009 == e.drawing


def test_delete_block(blocks, ac1009):
    archtick = blocks['_ARCHTICK']
    entities = list(archtick)
    archtick_name = archtick.name
    blocks.delete_block(archtick_name)
    assert archtick_name not in blocks
    db = ac1009.entitydb
    assert archtick._block_handle not in db
    assert archtick._endblk_handle not in db
    for entity in entities:
        assert entity.dxf.handle not in db


def test_safe_delete_block(blocks, ac1009):
    # block names are case insensitive
    msp = ac1009.modelspace()
    msp.add_blockref('_archtick', insert=(0, 0))
    with pytest.raises(ezdxf.DXFBlockInUseError):
        blocks.delete_block('_ArchTick', safe=True)


def test_delete_all_blocks(blocks):
    blocks.delete_all_blocks(safe=False)
    blocks = list(blocks)
    # assure not deleting layout blocks
    assert 2 == len(blocks)
    block_names = [block.name for block in blocks]
    block_names.sort()
    assert ['$MODEL_SPACE', '$PAPER_SPACE'] == block_names


def test_rename_block(blocks):
    block = blocks.new('RENAME_ME')
    assert block in blocks

    blocks.rename_block('RENAME_ME', 'NEW_NAME')
    assert 'NEW_NAME' in blocks

    # block names are case insensitive
    blocks.rename_block('New_Name', 'check_lower_case')
    assert 'Check_Lower_Case' in blocks

    # but originals name is preserved
    assert blocks['Check_Lower_Case'].name == 'check_lower_case'


@pytest.fixture(scope='module')
def ac1015():
    return ezdxf.new('AC1015')


@pytest.fixture
def ac1015_blocks(ac1015):
    if 'TestBlock' not in ac1015.blocks:
        block = ac1015.blocks.new('TestBlock')
        block.add_line((0, 0), (10, 10))
        block.add_line((0, 0), (10, 10))
        block.add_line((0, 0), (10, 10))
    return ac1015.blocks


def test_ac1015_dxf_block_structure(ac1015_blocks, ac1015):
    assert 'TestBlock' in ac1015_blocks
    block = ac1015_blocks['TestBlock']
    block_record_handle = block.block_record_handle

    # exists an associated block record entry?
    block_record = ac1015.sections.tables.block_records.get(block.name)
    assert block_record_handle == block_record.dxf.handle
    assert block_record.dxf.name == block.name


def test_ac1015_delete_block(ac1015_blocks, ac1015):
    block = ac1015_blocks['TestBlock']
    block_name = block.name
    entities = list(block)
    block_record_handle = block.block_record_handle

    block_count = len(ac1015_blocks)
    ac1015_blocks.delete_block(block_name)

    # removed from blocks load_section?
    assert block_count-1 == len(ac1015_blocks)
    assert block_name not in ac1015_blocks

    # all block related management data deleted?
    db = ac1015.entitydb
    assert block._block_handle not in db
    assert block._endblk_handle not in db
    assert block_record_handle not in db

    # removed from block records table?
    assert block_name not in ac1015.sections.tables.block_records

    # all entities deleted ?
    for entity in entities:
        assert entity.dxf.handle not in db
    # we are done!


def test_ac1015_delete_all_blocks(ac1015_blocks):
    ac1015_blocks.delete_all_blocks()
    blocks = list(ac1015_blocks)

    # assure not deleting layout blocks
    assert 2 == len(blocks)

    block_names = [block.name for block in blocks]
    block_names.sort()
    assert ['*Model_Space', '*Paper_Space'], block_names


def test_ac1015_rename_block(ac1015_blocks):
    block = ac1015_blocks.new('RENAME_ME')
    assert block in ac1015_blocks

    ac1015_blocks.rename_block('RENAME_ME', 'NEW_NAME')
    assert 'NEW_NAME' in ac1015_blocks


EMPTYSEC = """  0
SECTION
  2
BLOCKS
  0
ENDSEC
"""

TESTBLOCKS = """  0
SECTION
  2
BLOCKS
  0
BLOCK
  8
0
  2
$MODEL_SPACE
 70
     0
 10
0.0
 20
0.0
 30
0.0
  3
$MODEL_SPACE
  1

  0
ENDBLK
  5
21
  8
0
  0
BLOCK
 67
     1
  8
0
  2
$PAPER_SPACE
 70
     0
 10
0.0
 20
0.0
 30
0.0
  3
$PAPER_SPACE
  1

  0
ENDBLK
  5
5B
 67
     1
  8
0
  0
BLOCK
  8
0
  2
_ARCHTICK
 70
     0
 10
0.0
 20
0.0
 30
0.0
  3
_ARCHTICK
  1

  0
POLYLINE
  5
239
  8
0
  6
BYBLOCK
 62
     0
 66
     1
 10
0.0
 20
0.0
 30
0.0
 40
0.15
 41
0.15
  0
VERTEX
  5
403
  8
0
  6
BYBLOCK
 62
     0
 10
-0.5
 20
-0.5
 30
0.0
  0
VERTEX
  5
404
  8
0
  6
BYBLOCK
 62
     0
 10
0.5
 20
0.5
 30
0.0
  0
SEQEND
  5
405
  8
0
  6
BYBLOCK
 62
     0
  0
ENDBLK
  5
23B
  8
0
  0
BLOCK
  8
0
  2
_OPEN30
 70
     0
 10
0.0
 20
0.0
 30
0.0
  3
_OPEN30
  1

  0
LINE
  5
23D
  8
0
  6
BYBLOCK
 62
     0
 10
-1.0
 20
0.26794919
 30
0.0
 11
0.0
 21
0.0
 31
0.0
  0
LINE
  5
23E
  8
0
  6
BYBLOCK
 62
     0
 10
0.0
 20
0.0
 30
0.0
 11
-1.0
 21
-0.26794919
 31
0.0
  0
LINE
  5
23F
  8
0
  6
BYBLOCK
 62
     0
 10
0.0
 20
0.0
 30
0.0
 11
-1.0
 21
0.0
 31
0.0
  0
ENDBLK
  5
241
  8
0
  0
ENDSEC
"""


ezdxf/__init__.py
# Created: 10.03.2011
# Copyright (C) 2011-2018, Manfred Moitzi
# License: MIT License
import sys
if sys.version_info[:2] < (2, 7):
    raise ImportError("Package 'ezdxf' requires Python 2.7 or later!")

# name space imports - do not remove
from ezdxf.options import options  # example: ezdxf.options.template_dir = 'c:\templates'
from ezdxf.tools.importer import Importer
from ezdxf.tools import transparency2float, float2transparency  # convert transparency integer values to floats 0..1
from ezdxf.tools.rgb import int2rgb, rgb2int
from ezdxf.tools.pattern import PATTERN
from ezdxf.lldxf import const  # restore module structure ezdxf.const
from ezdxf.lldxf.validator import is_dxf_file, is_dxf_stream
from ezdxf.filemanagement import new, read, readfile, readzip
from ezdxf.tools.standards import setup_linetypes, setup_styles
from ezdxf.lldxf.const import DXFError  # base error exception
from ezdxf.lldxf.const import DXFStructureError, DXFVersionError, DXFTableEntryError, DXFAppDataError, DXFXDataError
from ezdxf.lldxf.const import DXFAttributeError, DXFValueError, DXFKeyError, DXFIndexError, DXFTypeError, DXFInvalidLayerName
from ezdxf.lldxf.const import DXFBlockInUseError
# name space imports - do not remove

import codecs
from ezdxf.lldxf.encoding import dxfbackslashreplace
codecs.register_error('dxfreplace', dxfbackslashreplace)  # setup DXF unicode encoder -> '\U+nnnn'
version = (0, 8, 10)  # also update VERSION in setup.py
VERSION = "0.8.10"
__version__ = VERSION
__author__ = "mozman <me@mozman.at>"


ezdxf/modern/oleframe.py
# Copyright (c) 2018 Manfred Moitzi
# License: MIT License
from __future__ import unicode_literals
from .graphics import none_subclass, entity_subclass, ModernGraphicEntity, DefSubclass, DXFAttr, DXFAttributes

oleframe_subclass = DefSubclass('AcDbOleFrame', {
    'version': DXFAttr(70),  # OLE version number
    'length': DXFAttr(90),  # Length of binary data
    # 310: Binary data (multiple lines)
    # 1: End of OLE data (the string "OLE")
})


class OLEFrame(ModernGraphicEntity):
    __slots__ = ()
    DXFATTRIBS = DXFAttributes(none_subclass, entity_subclass, oleframe_subclass)


ole2frame_subclass = DefSubclass('AcDbOle2Frame', {
    'version': DXFAttr(70),  # OLE version number
    'type': DXFAttr(3),  # content type as string e.g. "Paintbrush Picture"
    'upper_left_corner': DXFAttr(10, xtype='Point3D'),  # Upper-left corner (WCS)
    'lower_right_corner': DXFAttr(11, xtype='Point3D'),  # Upper-left corner (WCS)
    'ole_object_type': DXFAttr(71),  # OLE object type, 1 = Link; 2 = Embedded; 3 = Static
    'ole_tile_mode': DXFAttr(72),  # Tile mode descriptor: 0 = Object resides in model space; 1 = Object resides in paper space
    'length': DXFAttr(90),  # Length of binary data
    # 310: Binary data (multiple lines)
    # 1: End of OLE data (the string "OLE")
})


class OLE2Frame(ModernGraphicEntity):
    __slots__ = ()
    DXFATTRIBS = DXFAttributes(none_subclass, entity_subclass, ole2frame_subclass)


ezdxf/modern/lwpolyline.py
# Created: 25.03.2011
# Copyright (c) 2011-2018, Manfred Moitzi
# License: MIT License
from __future__ import unicode_literals
from contextlib import contextmanager
import array

from ezdxf.lldxf.types import DXFTag, DXFVertex
from ezdxf.lldxf.packedtags import VertexArray, replace_tags
from ezdxf.lldxf import loader

from .graphics import ExtendedTags, DXFAttr, DefSubclass, DXFAttributes
from .graphics import none_subclass, entity_subclass, ModernGraphicEntity

FORMAT_CODES = frozenset('xysebv')
DEFAULT_FORMAT = 'xyseb'

_LWPOLYLINE_TPL = """0
LWPOLYLINE
5
0
330
0
100
AcDbEntity
8
0
100
AcDbPolyline
90
0
70
0
43
0.0
"""

# Order doesn't matter, not valid for AutoCAD:
# If tag 90 is not the first TAG, AutoCAD does not close the polyline when the `close` flag is set.
lwpolyline_subclass = DefSubclass('AcDbPolyline', {
    'count': DXFAttr(90, xtype='Callback', getter='__len__', setter='_set_count'),
    # always return actual length and set tag 90
    '_count': DXFAttr(90),  # special attribute to update length
    'elevation': DXFAttr(38, default=0.0),
    'thickness': DXFAttr(39, default=0.0),
    'flags': DXFAttr(70, default=0),
    'const_width': DXFAttr(43, default=0.0),
    'extrusion': DXFAttr(210, xtype='Point3D', default=(0.0, 0.0, 1.0)),
})

LWPOINTCODES = (10, 20, 40, 41, 42)


class LWPolylinePoints(VertexArray):
    code = -10  # compatible with DXFTag.code
    VERTEX_CODE = 10
    START_WIDTH_CODE = 40
    END_WIDTH_CODE = 41
    BULGE_CODE = 42
    VERTEX_SIZE = 5
    __slots__ = ('value', )

    @classmethod
    def from_tags(cls, tags):
        """
        Setup point array from extended tags.

        Args:
            tags: ExtendedTags() object

        """
        subclass = tags.get_subclass('AcDbPolyline')

        def get_vertex():
            point.append(attribs.get(cls.START_WIDTH_CODE, 0))
            point.append(attribs.get(cls.END_WIDTH_CODE, 0))
            point.append(attribs.get(cls.BULGE_CODE, 0))
            return tuple(point)
        data = []
        point = None
        attribs = {}
        for tag in subclass:
            if tag.code in LWPOINTCODES:
                if tag.code == 10:
                    if point is not None:
                        data.extend(get_vertex())
                    point = list(tag.value[0:2])  # just use x, y coordinates, z is invalid but you never know!
                    attribs = {}
                else:
                    attribs[tag.code] = tag.value
        if point is not None:
            data.extend(get_vertex())
        return cls(data=data)

    def append(self, point, format=DEFAULT_FORMAT):
        super(LWPolylinePoints, self).append(compile_array(point, format=format))

    def dxftags(self):
        for point in self:
            x, y, start_width, end_width, bulge = point
            yield DXFVertex(self.VERTEX_CODE, (x, y))
            if start_width:
                yield DXFTag(self.START_WIDTH_CODE, start_width)
            if end_width:
                yield DXFTag(self.END_WIDTH_CODE, end_width)
            if bulge:
                yield DXFTag(self.BULGE_CODE, bulge)


@loader.register('LWPOLYLINE', legacy=False)
def tag_processor(tags):
    points = LWPolylinePoints.from_tags(tags)
    subclass = tags.get_subclass('AcDbPolyline')
    replace_tags(subclass, codes=LWPOINTCODES, packed_data=points)
    return tags


class LWPolyline(ModernGraphicEntity):
    __slots__ = ()
    TEMPLATE = tag_processor(ExtendedTags.from_text(_LWPOLYLINE_TPL))
    DXFATTRIBS = DXFAttributes(none_subclass, entity_subclass, lwpolyline_subclass)
    CLOSED = 1
    PLINEGEN = 128

    @property
    def AcDbPolyline(self):
        return self.tags.subclasses[2]

    @property
    def closed(self):
        return self.get_flag_state(self.CLOSED, name='flags')

    @property
    def lwpoints(self):
        return self.AcDbPolyline.get_first_tag(LWPolylinePoints.code)

    @closed.setter
    def closed(self, status):
        self.set_flag_state(self.CLOSED, status, name='flags')

    def __len__(self):
        return len(self.lwpoints)

    def __iter__(self):
        """
        Yielding tuples of (x, y, start_width, end_width, bulge).

        """
        return iter(self.lwpoints)

    def __getitem__(self, index):
        """
        Returns polyline point at position index as (x, y, start_width, end_width, bulge) tuple.

        """
        return self.lwpoints[index]

    def __setitem__(self, index, value):
        """
        Set polyline point at position index. Point format is fixed as 'xyseb'.

        Args:
            index: point index
            value: point value as (x, y, [start_width, [end_width, [bulge]]]) tuple

        """
        self.lwpoints[index] = compile_array(value)

    def __delitem__(self, index):
        del self.lwpoints[index]
        self.update_count()

    def vertices(self):
        """
        Yields all points as (x, y) tuples.

        """
        for point in self:
            yield point[0], point[1]

    def vertices_in_wcs(self):
        """
        Yields all points as (x, y, z) tuples in WCS.

        """
        ocs = self.ocs()
        elevation = self.get_dxf_attrib('elevation', default=0.)
        for x, y in self.vertices():
            yield ocs.to_wcs((x, y, elevation))

    def append(self, point, format=DEFAULT_FORMAT):
        """
        Append point to polyline, format specifies a user defined point format.

        Args:
            point: (x, y, [start_width, [end_width, [bulge]]]) tuple
            format: format string, default is 'xyseb'
                x = x coordinate
                y = y coordinate
                s = start width
                e = end width
                b = bulge value
                v = (x, y) as tuple

        """
        self.lwpoints.append(point, format=format)
        self.update_count()

    def insert(self, pos, point, format=DEFAULT_FORMAT):
        """
        Insert new point in front of positions pos, format specifies a user defined point format.

        Args:
            pos: insert position
            point: point data
            format: format string, default is 'xyseb'
                x = x coordinate
                y = y coordinate
                s = start width
                e = end width
                b = bulge value
                v = (x, y) as tuple

        """
        data = compile_array(point, format=format)
        self.lwpoints.insert(pos, data)
        self.update_count()

    def append_points(self, points, format=DEFAULT_FORMAT):
        """
        Append new points to polyline, format specifies a user defined point format.

        Args:
            points: iterable of point, point is (x, y, [start_width, [end_width, [bulge]]]) tuple
            format: format string, default is 'xyseb'
                x = x coordinate
                y = y coordinate
                s = start width
                e = end width
                b = bulge value
                v = (x, y) as tuple

        """
        for point in points:
            self.lwpoints.append(point, format=format)
        self.update_count()

    @contextmanager
    def points(self, format=DEFAULT_FORMAT):
        points = self.get_points(format=format)
        yield points
        self.set_points(points, format=format)

    def get_points(self, format=DEFAULT_FORMAT):
        """
        Returns all points as list of tuples, format specifies a user defined point format.

        Args:
            format: format string, default is 'xyseb'
                x = x coordinate
                y = y coordinate
                s = start width
                e = end width
                b = bulge value
                v = (x, y) as tuple

        """
        return [format_point(p, format=format) for p in self.lwpoints]

    def set_points(self, points, format=DEFAULT_FORMAT):
        """
        Remove all points and append new points.

        Args:
            points: iterable of point, point is (x, y, [start_width, [end_width, [bulge]]]) tuple
            format: format string, default is 'xyseb'
                x = x coordinate
                y = y coordinate
                s = start width
                e = end width
                b = bulge value
                v = (x, y) as tuple

        """
        self.lwpoints.clear()
        self.append_points(points, format=format)

    def clear(self):
        self.lwpoints.clear()
        self.update_count()

    def _set_count(self, value):
        # use special DXF attribute, because 'count' is a callback attribute
        self.set_dxf_attrib('_count', value)

    def update_count(self):
        self._set_count(len(self.lwpoints))


def format_point(point, format='xyseb'):
    """
    Reformat point components.

    Args:
        point: list or tuple of (x, y, start_width, end_width, bulge)
        format: format string, default is 'xyseb'
            x = x coordinate
            y = y coordinate
            s = start width
            e = end width
            b = bulge value
            v = (x, y) as tuple

    Returns: tuple of selected components

    """
    x, y, s, e, b = point
    v = (x, y)
    vars = locals()
    return tuple(vars[code] for code in format.lower() if code in FORMAT_CODES)


def compile_array(data, format='xyseb'):
    """
    Gather point components from input data.

    Args:
        data: list or tuple of point components
        format: format string, default is 'xyseb'
            x = x coordinate
            y = y coordinate
            s = start width
            e = end width
            b = bulge value
            v = (x, y) as tuple

    Returns: array.array('d', (x, y, start_width, end_width, bulge))

    """
    a = array.array('d', (0., 0., 0., 0., 0.))
    format = [code for code in format.lower() if code in FORMAT_CODES]
    for code, value in zip(format, data):
        if code not in FORMAT_CODES:
            continue
        if code == 'v':
            a[0] = value[0]
            a[1] = value[1]
        else:
            a['xyseb'.index(code)] = value
    return a


